### PR TITLE
Add CoreML ML Program Resize

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

Add CoreML ML Program Resize
- refactor existing logic to try and simplify and share between NeuralNetwork and MLProgram checks
- add handling for some new attributes
  - antialias and axes - should have been done when setting the CoreML EP max opset to 21

### Motivation and Context

Support priority models

### CODE REVIEW

File: onnxruntime/core/providers/coreml/builders/helper.cc [hunk 1]
```diff
@@ -84,16 +84,10 @@ bool IsInputSupported(const Node& node, const NodeArg& input,
       return false;
     }
 
-    if (dim == 0) {
-      if (node.OpType() == "Resize" && &input == node.InputDefs()[1]) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/helper.cc:88]: To be really correct this also needed to allow an empty input for `scales` if opset was 11. That was getting too convoluted for here so added a simple flag so an op builder implementation can allow empty inputs in this basic check and handle them internally.

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/resize_op_builder.cc [hunk 2]
```diff
@@ -251,14 +251,37 @@ bool ResizeOpBuilder::IsOpSupportedImpl(const GraphViewer& graph_viewer, const N
...
                               << ", scale_n, " << scale_n << ", scale_c, " << scale_c;
         return false;
       }
+
+      // if downsampling the input size must be evenly divisible by the output size to match the onnx output
+      if (scale_h < 1.0f || scale_w < 1.0f) {
+        // we also require input_shape to be known to check
+        auto h_in = input_is_nchw ? input_shape[2] : input_shape[1];
+        auto w_in = input_is_nchw ? input_shape[3] : input_shape[2];
+        if (h_in == 0 || w_in == 0) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/resize_op_builder.cc:271]: note: NNAPI EP implementation uses 0 for an unknown dim to match NNAPI (vs. the value being -1 in the CoreML and XNNPACK EPs)

File: onnxruntime/core/providers/coreml/builders/impl/resize_op_builder.cc [hunk 3]
```diff
@@ -228,33 +420,59 @@ bool ResizeOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputPa
...
-        LOGS(logger, VERBOSE) << "Input shape with dynamic dimensions is not supported.";
+    } else if (scale_h <= 1.f && scale_w <= 1.f) {
+      // downsample
+      if (input_params.create_mlprogram) {
+        auto h_in = input_shape[input_rank - 2];
+        auto w_in = input_shape[input_rank - 1];
+        auto h_out = h_in * scale_h;
+        auto w_out = w_in * scale_w;
+
+        if (std::fmod(float(h_in), h_out) != 0.f) {
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/providers/coreml/builders/impl/resize_op_builder.cc]: Great catch! Refined the approach.

ML Program can, but it's a little touchy around how many '3's in 0.333... are required. I don't think that is relevant now as I'm using the output sizes with resize_bilinear to run downsampling due to other flakiness in the output from upsample_bilinear. local tests are fine but CIs with different macOS versions produce invalid output. resize_bilinear is stable and matches onnx.

File: onnxruntime/core/providers/coreml/builders/impl/resize_op_builder.cc [hunk 4]
```diff
@@ -423,21 +424,17 @@ bool ResizeOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputPa
...
         auto h_in = input_shape[input_rank - 2];
         auto w_in = input_shape[input_rank - 1];
-        auto h_out = double(h_in) * scale_h;
-        auto w_out = double(w_in) * scale_w;
 
-        if (std::floor(h_out) != h_out) {
-          LOGS(logger, VERBOSE) << "Resize: downsampling output height: " << h_out
+        if (!utils::IsScaleDividingByFactorOfN(h_in, scale_h)) {
+          LOGS(logger, VERBOSE) << "Resize: downsampling scale " << scale_h
                                 << " is not a factor of input height: " << h_in;
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/impl/resize_op_builder.cc:432]: pedantically, "is not dividing by a factor of input height" may be more precise but it's probably fine

**Review by edgchen1 — DISMISSED** [review]

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**